### PR TITLE
Fixed infinite quadrant subdivision

### PR DESCRIPTION
**Pull request by w8r**
This fix addresses following problem: if there are many nodes and they are positioned pretty densely, Barnes-Hut quadtree can easily run out of number precision and go into infinite loop while falling into the same quadrant. There's no other solution for that but put a limit on subdivision, for which I arbitrarily chose 3 attempts.

I added a test case for it. But I don't know how to formalize it better, since if you remove my patch from the code, the whole test suite will crash, test would not even throw. In the browser it will crash the tab completely, no debugging possible.

Also the quadrants are square now, which is supposed to be aimed at the cases when nodes are all lined up vertically or horizontally.

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -8,6 +8,8 @@ var assert = require('assert'),
 var helpers = require('./helpers.js'),
     layout = require('./index.js');
 
+var clusters = require('graphology-generators').random.clusters;
```
Comment (Yomguithereal) [inline test.js]: This should be `require('graphology-generators/random/clusters')` to save loading time.

**Comment by Yomguithereal** [comment]
Hello @w8r. Thanks for noticing this bug. May I ask how you stumbled upon it? We use the algorithm since a long time on very different graphs and never fell on it.

> I added a test case for it. But I don't know how to formalize it better, since if you remove my patch from the code, the whole test suite will crash, test would not even throw. In the browser it will crash the tab completely, no debugging possible.

I guess one cannot do better. I don't see a way to test if something runs an infinite loop or not since the synchronous execution would delay timeouts and such.

> Also the quadrants are square now, which is supposed to be aimed at the cases when nodes are all lined up vertically or horizontally.

This might be an issue. I am not sure we want that.

@jacomyma, what's your opinion about all this?

File: test.js [hunk 2]
```diff
@@ -154,4 +156,27 @@ describe('graphology-layout-forceatlas2', function() {
       }, /linLogMode/);
     });
   });
+
+  describe('Barnes-Hut optimization', function() {
+    it('should converge on a large random graph with small coordinate values', function() {
+      // Creating a random clustered graph
+      var graph = clusters(Graph, {
```
Comment (Yomguithereal) [inline test.js]: You should provide a `rng` option to the generator using the `seedrandom` library so the test remains deterministic.

**Comment by w8r** [comment]
@Yomguithereal well, as I mentioned, on big graphs with small coordinate values it resulted in fatal crashes repeatedly and predictably, so I had no other way as to debug and find out.

Another thing is that the algorithm doesn't work at all, if all the nodes are stacked on `[0,0]` position, but I suppose you know about it already.

**Comment by Yomguithereal** [comment]
Can you open a separate PR with the squaring of quadrants? I spoke with @jacomyma and it seems it does not cause any issues on our side.

**Comment by w8r** [comment]
#3

**Comment by Yomguithereal** [comment]
Thanks @w8r. Will review #3 with @jacomyma beginning of next week.

**Comment by w8r** [comment]
Thanks!

**Comment by Yomguithereal** [comment]
Thanks @w8r. I merged and ran some lint etc. Waiting for #3 next week to publish an update on npm (or tell me if you absolutely need one before and I can publish one earlier).

**Comment by w8r** [comment]
Thanks a lot! No, for me it's no rush, I'm not using graphology, but only the ForceAtlas. I found the bug months ago and just felt like contributing the fix that I've been using.

**Comment by Yomguithereal** [comment]
Are you using the iteration function with a FloatArray you prepare before or are you using another implementation altogether?

**Comment by w8r** [comment]
I am using the implementation from sygma.js. This one needs patching as well

**Pull request merged**